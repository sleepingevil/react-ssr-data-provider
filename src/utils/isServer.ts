const IS_SERVER = typeof window === 'undefined';

export const isServer = () => IS_SERVER;
