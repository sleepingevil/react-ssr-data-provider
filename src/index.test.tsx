import { GetDataProviderResult, DataContext, getDataProvider, useDataEffect } from ".";
import React, { useState, useEffect } from 'react';
import { render, RenderResult, act } from "@testing-library/react";
import { isServer } from "./utils/isServer";

jest.mock('./utils/isServer', () => ({
  isServer: jest.fn()
}));

describe('DataProvider', () => {
  let provider: GetDataProviderResult<{ getData: () => Promise<string> }>;

  describe('When on the server', () => {
    beforeEach(() => {
      (isServer as jest.Mock).mockReturnValue(true);
      provider = getDataProvider<{ getData: () => Promise<string> }>();
    });

    test('it should provide the getScript method that returns the script to inject pre-fetched data', async () => {
      const dataPromise = Promise.resolve('someData');
      const TestComponent: React.FC = () => {
        const { getData } = provider.useDataClient();
        const [data, setData] = useState<string>();

        useDataEffect(() => {
          const doEffect = async () => {
            setData(await getData())
          };

          doEffect();
        }, [])


        return <div>{data}</div>;
      };

      const context = {} as DataContext;

      const renderResult: RenderResult = render(<provider.DataProvider providers={{ getData: () => dataPromise }} dataContext={context}>
        <TestComponent />
      </provider.DataProvider>, {});

      expect(context).toHaveProperty('getScript', expect.any(Function));

      await act(async () => {
        expect(await context.getScript!()).toMatchSnapshot();
      });

      expect(renderResult.baseElement).toMatchSnapshot();
    });

    test('it should provide the getScript method that returns an empty script when no data was requested', async () => {
      const dataPromise = Promise.resolve('someData');
      const TestComponent: React.FC = () => {
        return <div></div>;
      };

      const context = {} as DataContext;

      const renderResult: RenderResult = render(<provider.DataProvider providers={{ getData: () => dataPromise }} dataContext={context}>
        <TestComponent />
      </provider.DataProvider>, {});

      expect(context).toHaveProperty('getScript', expect.any(Function));

      await act(async () => {
        expect(await context.getScript!()).toMatchSnapshot();
      });

      expect(renderResult.baseElement).toMatchSnapshot();
    });

    test('it should render without a context passed in', async () => {
      const dataPromise = Promise.resolve('someData');
      const TestComponent: React.FC = () => {
        return <div></div>;
      };

      const renderResult: RenderResult = render(<provider.DataProvider providers={{ getData: () => dataPromise }}>
        <TestComponent />
      </provider.DataProvider>, {});


      await act(async () => {
        await dataPromise;
      });

      expect(renderResult.baseElement).toMatchSnapshot();
    });
  });

  describe('When in the browser', () => {
    beforeEach(() => {
      (isServer as jest.Mock).mockReturnValue(false);
    });

    test('it should render consistently', async () => {
      provider = getDataProvider<{ getData: () => Promise<string> }>();
      const dataPromise = Promise.resolve('someData');
      const TestComponent: React.FC = () => {
        const { getData } = provider.useDataClient();
        const [data, setData] = useState<string>();

        useDataEffect(() => {
          const doEffect = async () => {
            setData(await getData())
          };

          doEffect();
        }, [])


        return <div>{data}</div>;
      };

      const renderResult = render(<provider.DataProvider providers={{ getData: () => dataPromise }}>
        <TestComponent />
      </provider.DataProvider>, {});

      await act(async () => {
        await dataPromise;
      });

      expect(renderResult.baseElement).toMatchSnapshot();
    });

    test('it should render consistently when state is pre-populated', async () => {
      (window as any).__dataProvider__ = {
        prefetchedState: {
          getData: 'someCachedData'
        }
      };

      provider = getDataProvider<{ getData: () => Promise<string> }>();

      const dataPromise = Promise.resolve('someData');
      const TestComponent: React.FC = () => {
        const { getData } = provider.useDataClient();
        const [data, setData] = useState<string>();

        useDataEffect(() => {
          const doEffect = async () => {
            setData(await getData())
          };

          doEffect();
        }, [])


        return <div>{data}</div>;
      };

      let renderResult: RenderResult;

      await act(async () => {
        renderResult = render(<provider.DataProvider providers={{ getData: () => dataPromise }}>
          <TestComponent />
        </provider.DataProvider>);
        await dataPromise;
      });

      expect(renderResult.baseElement).toMatchSnapshot();

      delete (window as any).__dataProvider__;
    });

    test('it should only use the cache once', async () => {
      (window as any).__dataProvider__ = {
        prefetchedState: {
          getData: 'someCachedData'
        }
      };

      provider = getDataProvider<{ getData: () => Promise<string> }>();

      const dataPromise = Promise.resolve('someData');
      const TestComponent: React.FC<{ shouldUpdate?: boolean }> = ({ shouldUpdate }) => {
        const { getData } = provider.useDataClient();
        const [updater, setUpdater] = useState(Math.random());
        const [data, setData] = useState<string>();

        useEffect(() => {
          if (shouldUpdate) {
            setUpdater(Math.random());
          }
        }, [shouldUpdate]);

        useDataEffect(() => {
          const doEffect = async () => {
            setData(await getData())
          };

          doEffect();
        }, [updater])


        return <div>{data}</div>;
      };

      let renderResult: RenderResult;

      await act(async () => {
        renderResult = render(<provider.DataProvider providers={{ getData: () => dataPromise }}>
          <TestComponent shouldUpdate />
        </provider.DataProvider>);
      });

      await act(async () => {
        renderResult.rerender(<provider.DataProvider providers={{ getData: () => dataPromise }}>
          <TestComponent shouldUpdate />
        </provider.DataProvider>);
        await dataPromise;
      });


      expect(renderResult.baseElement).toMatchSnapshot();

      delete (window as any).__dataProvider__;
    });
  });
})
