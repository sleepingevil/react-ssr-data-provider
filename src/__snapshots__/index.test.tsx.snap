// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataProvider When in the browser it should only use the cache once 1`] = `
<body>
  <div>
    <div>
      someData
    </div>
  </div>
</body>
`;

exports[`DataProvider When in the browser it should render consistently 1`] = `
<body>
  <div>
    <div>
      someData
    </div>
  </div>
</body>
`;

exports[`DataProvider When in the browser it should render consistently when state is pre-populated 1`] = `
<body>
  <div>
    <div>
      someCachedData
    </div>
  </div>
</body>
`;

exports[`DataProvider When on the server it should provide the getScript method that returns an empty script when no data was requested 1`] = `""`;

exports[`DataProvider When on the server it should provide the getScript method that returns an empty script when no data was requested 2`] = `
<body>
  <div>
    <div />
  </div>
</body>
`;

exports[`DataProvider When on the server it should provide the getScript method that returns the script to inject pre-fetched data 1`] = `"<script>window.__dataProvider__ = {prefetchedState: {\\"0\\":{\\"getData\\":\\"someData\\"}}}</script>"`;

exports[`DataProvider When on the server it should provide the getScript method that returns the script to inject pre-fetched data 2`] = `
<body>
  <div>
    <div />
  </div>
</body>
`;

exports[`DataProvider When on the server it should render without a context passed in 1`] = `
<body>
  <div>
    <div />
  </div>
</body>
`;
